File: ios/PwNode.swift
import Foundation

struct PwNode: Identifiable {
    let id = UUID()
    let url: URL
    let children: [PwNode]?

    var name: String {
         return url.deletingPathExtension().lastPathComponent
    }

    var isLeaf: Bool {
        return (children ?? []).isEmpty
    }

    static func loadFrom(_ fromDir: URL) throws -> Self {
        var children: [Self] = []

        for url in try FileManager.default.ls(fromDir) {
            let node = FileManager.default.isDir(url) ?
                                try loadFrom(url) :
                                PwNode(url: url, children: nil)

            children.append(node)
        }

        return PwNode(url: fromDir, children: children)
    }

    /// Returns a subset of the tree with paths to every node that matches
    /// `predicate`.
    func findChildren(predicate: String) -> [PwNode] {
        var matches: [PwNode] = []
        let predicate = predicate.lowercased()

        for child in children ?? [] {
            let childMatches = child.findChildren(predicate: predicate)

            if childMatches.isEmpty {
                // Append the child with all its children if it matches the
                // predicate.
                if child.name.lowercased().contains(predicate) {
                    matches.append(child)
                }
            } else {
                // Append the child with a subset of its own children if one of
                // its children matched the predicate.
                let subsetChild = PwNode(url: child.url, children: childMatches)
                matches.append(subsetChild)
            }
        }

        return matches
    }

    func getPlaintext() -> String? {
        if !self.isLeaf {
            return nil
        }
        return Age.decrypt(self.url)
    }
}

File: android/src/main/kotlin/one/kafva/kage/types/PwNode.kt
package one.kafva.kage.types

import java.io.File

data class PwNode(
    private val path: File,
    var children: List<PwNode>,
) {
    companion object {
        fun fromRoutePath(serialisedNodePath: String): String =
            serialisedNodePath.replace("|", "/")

        fun prettyName(nodePath: String): String =
            nodePath.split("/").last().removeSuffix(".age")
    }

    val name = path.getName().removeSuffix(".age")
    private val isPassword = path.getName().endsWith(".age")
    private val pathString = path.toPath().toString()

    fun toRoutePath(filesDir: File): String =
        pathString.removePrefix("${filesDir.path}/").replace("/", "|")

    /**
     * Recursively load the nodes under the current path, needs to be called
     * explicitly after creating an object.
     */
    fun loadChildren() {
        if (!path.isDirectory()) {
            return
        }

        val mutableChildren: MutableList<PwNode> = mutableListOf()

        path.listFiles()?.forEach {
            val name = it.getName() ?: ""
            if (name != "" && !name.startsWith(".")) {
                val child = PwNode(it, listOf())
                child.loadChildren()
                mutableChildren.add(child)
            }
        }

        children = mutableChildren.toList()
    }

    /** Predicate should be provided in lowercase! */
    fun findChildren(predicate: String): List<PwNode> {
        // Include all of the children if there is no query
        if (predicate.isEmpty()) {
            return children
        }

        val matches: MutableList<PwNode> = mutableListOf()

        // If the parent does not match the predicate, check each child
        for (child in children) {
            if (child.name.lowercase().contains(predicate)) {
                // Include everything beneath a matching child
                matches.add(child)
                continue
            }

            // Check children recursively if the current child was not a match
            val childMatches = child.findChildren(predicate)

            // Include the child with the subset of the child nodes that match the query
            if (childMatches.isNotEmpty()) {
                val subsetChild = PwNode(child.path, childMatches.toList())
                matches.add(subsetChild)
            }
        }

        return matches.toList()
    }

    /** Recursive count of children */
    fun passwordCount(): Int {
        val childCount = children.count { it.isPassword }
        return childCount + children.sumOf { it.passwordCount() }
    }
}
